interface/modules/zend_modules/config/application.config.php

```diff
@@ -1,24 +1,43 @@
 <?php
+/**
+ *  You can see the default application configuration options here
+ *  as well as how the individual module.config.php files should operate.
+ *  @see https://docs.zendframework.com/zend-mvc/services/
+ *
+ */
+// The required application modules we need to load all the time are listed here.
+$core_modules = [
+    'Zend\Router', // this was separated from the MVC module into it's own module.  Handles the routing layers.
+    'Zend\Validator',  // Handles validations....
+    'Zend\Mvc\I18n', // TODO: since translations seem to be using the OPENEMR translations we may not need this anymore though we load it as a factory
+    'Zend\Form', // needed for the formHelper view plugin.
+    'Application', // Main application module starting point.
+    'Installer', // Handles the dynamic adding / removing of modules in the system.
+    'Acl', // Handles all of the permission checks in the system.
+    'Carecoordination', // Handles import / export of CCR,CDA Immunization, Syndromicsurveillance
+    'Ccr', // Module specific code for dealing with CCR import/export
+    'Documents', // Handles the loading / creating of documents
+    'Immunization',
+    'Syndromicsurveillance',
+    'Patientvalidation', // Validates patients for duplicate records
+    'Multipledb', // Allows multiple database handlers within the module system
+    'PrescriptionTemplates' // Handles the printing / displaying of prescriptions.
+];
```

Comment:
Do we still need to make this file writable?

Reply:
It should be locked down like the rest of the codebase.  Only developers will be the ones modifying it from now on so we can update the docker images to reflect that.

Reply:
This is great news and will simplify install (since no longer need write privileges)!

Reply:
Hi @adunsulag 
I have tested your code and you did very good work!
Regarding those changes in the application.config:
first, if this file is not writeable that mean the Installer module cannot install new modules, this was the main feature of the installer ( I saw in the first commit that you try to make it using a new file 'modules.config.php', why you decided to remove it?)
second, I don't think the idea that developers will modify the file is good, the integration of new modules need be without changes in the codebase it creates conflicts when you want to upgrade the code. 

You added a comment in the openemr/interface/modules/zend_modules/config/application.config.php:29
```
// TODO: if we ever want to load module definitions from the database we would do that there...
$plugin_modules = [];
```    
Maybe it's can be a good solution for that issues..

Reply:
@matrix-amiel I agree its a nice feature to have to add modules without modifying the codebase.  What I saw in working with the module loader is that the only modules that showed up were the ones already listed in the application.config.php file.  The installer would then attempt to add the module back in.  Since it already existed, the installer would just re-add the same entry and then write out the exact same file to the system.  

The installer will still 'install' the module by activating it and inserting any hooks / etc into the database.  So the modules will still get added / hooks configured  and you can connect them to the menu system.  I just removed the whole writeable file thing as it wasn't really doing anything as far as I could tell.

It didn't make sense to have a writeable file that just had the same contents written out to it every time.

You're team has written a lot of modules, does it do anything differently on your system?  Do you add your modules as composer dependencies in the vendor directory?  I thought maybe that was the only difference on how the modules got added but I didn't see anyone doing that.

We can pull additional modules from the database but the system will have to do the query outside of Zend.  The zend database adapters haven't been loaded yet so we can't use Zend queries to load the names of the modules.  We could use the OpenEMR sqlStatement functions to load the queries.  

I agree its nice to be able to add modules without modifying / committing to the codebase. 
 That has been my goal to be able to have a wordpress type of plugin install environment. 
 However, I didn't see how the current installer code was really accomplishing that as it looks like the module had to be in the application config file in order to install it (since Zend can't autoload it unless it was listed in the application config file) only to have it write out to the file once again.

@bradymiller mentioned it is an installer nightmare to have that file writeable.  
My concern is the performance impact of doing a database query on every page request to check if the modules are loaded.  Now we could cache these module definitions and store them in a shared memory store.  We could store it in the APC cache if people have it enabled, or in a redis / memcache database to improve performance, but I wasn't sure how many people had that available.

My goal for this round was to avoid any slow downs in the system but I might be getting caught into a pre-optimization trap.  It looks like for the 8 modules that we have, the average load / request execute time is about 71ms on my unoptimized 4 core laptop.  I can't imagine another db query will add that much more since we are already doing a number of queries (loading the globals, etc).  I'll give it a try at loading the modules names from the database and measure the performance, it might not be that bad.  I'll let you know what my results are.

Reply:
@matrix-amiel My latest update to this pulls the module definitions from the database.  I measured the performance and it's in the microseconds on a local machine, even on a cloud server should be fairly fast.  

So this should be resolved now.  Let me know if that addresses your concerns.  I'm not sure if end users would do it, but a great thing for developers to load additional modules into the folder structure without 'modifying' the codebase would be adding the composer/installers package to the package.  We can key off the zend-module tag and it would drop the module folder in the right directory.  https://packagist.org/packages/composer/installers

Reply:
@adunsulag 
I really love your suggestion. 
The remove a writeable from that file is an old dream...   (#2086)
If it will work well I think it's best.

We already use composer/installers (with extensions) to put the modules in the zend_modules/modules folder and install them in our project and it's a good option for developers.

The option you added to load the modules from vendor is very nice but note that Installer module will not find them.

Reply:
@matrix-amiel What are your thoughts on putting this into master?  Do we need to give your organization time to update your modules before we merge this in?  

I'd like to get in some of my additional changes that will show how the modules can better extend the codebase using the current EventDispatcher in OpenEMR.  But I won't be able to bring that code in until this has code has been merged into master.

As for loading / registering modules from the vendor directory, I think we could customize the Installer module to use Symfony's Finder component https://symfony.com/doc/current/components/finder.html to find the Modules so we wouldn't need the composer/installers piece.  I think it'd be a bit of work though so not sure it's worth the effort right now as to remove any false positives we'd probably have to open the composer.json file per project and verify it has the zend-module tag or attempt to load the class etc.